union.go

```diff
@@ -0,0 +1,58 @@
+package rwvfs
+
+import (
+	"io"
+	pathpkg "path"
+
+	"golang.org/x/tools/godoc/vfs"
+)
+
+// Union returns a new FileSystem which is the union of the provided file systems.
+func Union(fileSystems ...FileSystem) FileSystem {
+	if len(fileSystems) == 0 {
+		return Map(nil)
+	}
+	if len(fileSystems) == 1 {
+		return fileSystems[0]
+	}
+
+	ns := vfs.NameSpace{}
+	for _, fs := range fileSystems {
+		ns.Bind("/", fs, "/", vfs.BindAfter)
+	}
+	return &unionFs{
+		NameSpace:   ns,
+		fileSystems: fileSystems,
+	}
+}
+
+type unionFs struct {
+	vfs.NameSpace
+	fileSystems []FileSystem
+}
+
+func (fs *unionFs) resolve(path string) FileSystem {
+	parent := pathpkg.Dir(path)
+	for _, fs2 := range fs.fileSystems {
+		if _, err := fs2.Stat(parent); err == nil {
+			return fs2
+		}
+	}
+	return fs.fileSystems[0]
+}
+
+// Create creates the named file, truncating it if it already exists.
+func (fs *unionFs) Create(path string) (io.WriteCloser, error) {
+	return fs.resolve(path).Create(path)
```

Comment:
A lock should be held during both the calls to `fs.resolve(path)` and `(FileSystem).{Create,Mkdir,Remove}`. That's because the latter operation affects the result of the former, so you have [TOCTTOU](http://en.wikipedia.org/wiki/Time_of_check_to_time_of_use) race conditions here.

This might really hurt performance, unfortunately. I think that's highly likely. In that case, I think you can just:

* Try to think if there are any really bad bugs (panics, etc.) that could occur.
* If not, then just add a comment to the Union func and/or unionFS type explaining the race condition and why we're punting on solving it.
* If so, then we probably do need to fix it.